From 8ec4d4522fa6679f7c9daad9dab4f438071953ee Mon Sep 17 00:00:00 2001
From: Steven Liu <steven.liu@rock-chips.com>
Date: Wed, 14 Oct 2020 17:42:00 +0800
Subject: [PATCH] pwm: rockchip: Add pwm output center aligned mdoe

Support pwm output aligned mode to switch from  left-aligned
to center-aligned. In dts, add "center-aligned".

Signed-off-by: Steven Liu <steven.liu@rock-chips.com>
Change-Id: I3e699c873a9ef533e59e11dbf9777001f205b4d9
---
 drivers/pwm/pwm-rockchip.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/pwm/pwm-rockchip.c b/drivers/pwm/pwm-rockchip.c
index d81f4e4c3bf6..5263934c3b6b 100644
--- a/drivers/pwm/pwm-rockchip.c
+++ b/drivers/pwm/pwm-rockchip.c
@@ -30,6 +30,7 @@
 #define PWM_INACTIVE_POSITIVE	(1 << 4)
 #define PWM_POLARITY_MASK	(PWM_DUTY_POSITIVE | PWM_INACTIVE_POSITIVE)
 #define PWM_OUTPUT_LEFT		(0 << 5)
+#define PWM_OUTPUT_CENTER	(1 << 5)
 #define PWM_LOCK_EN		(1 << 6)
 #define PWM_LP_DISABLE		(0 << 8)
 
@@ -42,6 +43,7 @@ struct rockchip_pwm_chip {
 	const struct rockchip_pwm_data *data;
 	void __iomem *base;
 	bool vop_pwm_en; /* indicate voppwm mirror register state */
+	bool center_aligned;
 };
 
 struct rockchip_pwm_regs {
@@ -189,6 +191,11 @@ static int rockchip_pwm_enable(struct pwm_chip *chip,
 	val = readl_relaxed(pc->base + pc->data->regs.ctrl);
 	val &= ~pc->data->enable_conf_mask;
 
+	if (PWM_OUTPUT_CENTER & pc->data->enable_conf_mask) {
+		if (pc->center_aligned)
+			val |= PWM_OUTPUT_CENTER;
+	}
+
 	if (enable)
 		val |= enable_conf;
 	else
@@ -410,6 +417,9 @@ static int rockchip_pwm_probe(struct platform_device *pdev)
 		pc->chip.of_pwm_n_cells = 3;
 	}
 
+	pc->center_aligned =
+		device_property_read_bool(&pdev->dev, "center-aligned");
+
 	ret = pwmchip_add(&pc->chip);
 	if (ret < 0) {
 		clk_unprepare(pc->clk);
-- 
2.35.3

